<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
   <TITLE> [antlr-interest] AntLR C# target exceptions problem
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:antlr-interest%40antlr.org?Subject=Re:%20%5Bantlr-interest%5D%20AntLR%20C%23%20target%20exceptions%20problem&In-Reply-To=%3C49E3556B.2020405%40rwth-aachen.de%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="034031.html">
   <LINK REL="Next"  HREF="034035.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[antlr-interest] AntLR C# target exceptions problem</H1>
    <B>Jan Newger</B> 
    <A HREF="mailto:antlr-interest%40antlr.org?Subject=Re:%20%5Bantlr-interest%5D%20AntLR%20C%23%20target%20exceptions%20problem&In-Reply-To=%3C49E3556B.2020405%40rwth-aachen.de%3E"
       TITLE="[antlr-interest] AntLR C# target exceptions problem">Jan.Newger at rwth-aachen.de
       </A><BR>
    <I>Mon Apr 13 08:08:27 PDT 2009</I>
    <P><UL>
        <LI>Previous message: <A HREF="034031.html">[antlr-interest] AntLR C# target exceptions problem
</A></li>
        <LI>Next message: <A HREF="034035.html">[antlr-interest] AntLR C# target exceptions problem
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#34033">[ date ]</a>
              <a href="thread.html#34033">[ thread ]</a>
              <a href="subject.html#34033">[ subject ]</a>
              <a href="author.html#34033">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>Johannes Luber wrote:
&gt;&gt;<i> Johannes Luber wrote:
</I>&gt;&gt;&gt;<i> Does putting a breakpoint into the catch block work? Otherwise the menu
</I>&gt;&gt;<i> for exceptions offers the options to break when an exception is thrown and
</I>&gt;&gt;<i> if an exception is caught. Furthermore, did you add a messagebox call or
</I>&gt;&gt;<i> console print to check if the catch block is actually executed (in release
</I>&gt;&gt;<i> mode)? Otherwise I can't check things out myself, as my development machine
</I>&gt;&gt;<i> has a display problem and is thus unusable right now.
</I>&gt;&gt;<i> Putting a breakpoint into the catch block doesn't work. If you look at
</I>&gt;&gt;<i> the code I posted, there actually is a messagebox call, which is never
</I>&gt;&gt;<i> executed (debug or release mode doesn't matter).
</I>&gt;&gt;<i>
</I>&gt;&gt;<i> I noticed however, that when a NoViableAltException is thrown from my
</I>&gt;&gt;<i> generated parser, that the parser itself catches it, calls ReportError
</I>&gt;&gt;<i> then Recover and ultimately returns silently from this method (start
</I>&gt;&gt;<i> symbol of my grammar). So in fact the exception is swallowed by the
</I>&gt;&gt;<i> parser itself. Is this &quot;by design&quot;? I just didn't expect this kind of
</I>&gt;&gt;<i> behavior, because it seems to be quite different from the java runtime
</I>&gt;&gt;<i> approach. What would be the solution? Should I override the ReportError
</I>&gt;&gt;<i> method in my parser and rethrow the exception?
</I>&gt;<i> 
</I>&gt;<i> I've never used the error recovery process myself, but I think that you have to overwrite 3 methods. If ReportError() is one of them I don't know but &lt;<A HREF="http://www.antlr.org/wiki/display/ANTLR3/Error+reporting+and+recovery">http://www.antlr.org/wiki/display/ANTLR3/Error+reporting+and+recovery</A>&gt; has some insights. Don't forget to use .NET capitalisation instead the Java one.
</I>&gt;<i> 
</I>
Thanks for your quick response, Johannes. Actually, I don't want to
override the error reporting mechanism, I just want the runtime to throw
the respective exceptions when the input is malformed (just in the same
way the java runtime does it).
Is the behavior I previously described considered a bug or is this &quot;by
design&quot;? It seems odd to me that exceptions are swallowed within the
parser itself. So even if I where to replace the error reporting
mechanism, I would change the error message of an exception which never
reaches my code in the first place. IMHO this is ether a serious bug or
I'm doing something incredibly stupid here.

Cheers
Jan
</PRE>




















<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="034031.html">[antlr-interest] AntLR C# target exceptions problem
</A></li>
	<LI>Next message: <A HREF="034035.html">[antlr-interest] AntLR C# target exceptions problem
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#34033">[ date ]</a>
              <a href="thread.html#34033">[ thread ]</a>
              <a href="subject.html#34033">[ subject ]</a>
              <a href="author.html#34033">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="http://www.antlr.org/mailman/listinfo/antlr-interest">More information about the antlr-interest
mailing list</a><br>
</body></html>
